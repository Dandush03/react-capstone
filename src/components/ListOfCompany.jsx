import React, { Component } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import PropTypes from 'prop-types';

import { getCompaniesArray } from '../actions/fetchCompanies';
import { handlePopUp } from '../actions/popUp';

import Company from './Company';
import Filter from './Filter';

class Companies extends Component {
  constructor(props) {
    super(props);
    this.togglePopup = this.togglePopup.bind(this);
  }

  componentDidMount() {
    const { props: { getCompaniesArray } } = this;
    getCompaniesArray();
  }

  togglePopup(btn) {
    const { props: { handlePopUp } } = this;
    handlePopUp(btn, true);
  }

  render() {
    const { props, togglePopup } = this;
    const { companies, popUp, filter } = props;
    let showCompanies;
    if (filter === 'all') {
      showCompanies = companies;
    } else {
      showCompanies = companies.filter(c => (c.name && c.name.split('').shift().toLowerCase() === filter));
    }
    const companiesList = showCompanies.map(company => (
      <button type="button" key={`company-${company.symbol}`} className="company-list" onClick={() => togglePopup(company)}>
        <h1>{company.name}</h1>
        <span>{`(${company.symbol})`}</span>
      </button>
    ));
    return (
      <div className="companies">
        <Filter />
        {companiesList}
        {popUp.show ? <Company /> : null}
      </div>
    );
  }
}

Companies.propTypes = {
  handlePopUp: PropTypes.func.isRequired,
  getCompaniesArray: PropTypes.func.isRequired,
  companies: PropTypes.arrayOf(PropTypes.object).isRequired,
  popUp: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.bool, PropTypes.object])).isRequired,
  filter: PropTypes.string.isRequired,
};

const structeredSelector = createStructuredSelector({
  companies: state => state.companies,
  popUp: state => state.popUp,
  filter: state => state.filter,
});

const mapDispatchToProps = { getCompaniesArray, handlePopUp };

export default connect(structeredSelector, mapDispatchToProps)(Companies);
