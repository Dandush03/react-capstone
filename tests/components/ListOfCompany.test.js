import React from 'react';
import renderer from 'react-test-renderer';

import storeConfig from '../../src/store';
import ListOfCompany from '../../src/components/ListOfCompany';


const store = storeConfig();

describe('components', () => {
  describe('ListOfCompany', () => {
    const component = renderer.create(<ListOfCompany store={store} />);
    const tree = component.toJSON();
    it('Should match to Snapshot', () => {
      expect(tree).toMatchSnapshot();
    });
  });
});
