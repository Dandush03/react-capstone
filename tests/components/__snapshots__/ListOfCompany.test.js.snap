// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ListOfCompany Should match to Snapshot 1`] = `
<div
  className="companies"
>
  <button
    className="company-list"
    onClick={[Function]}
    type="button"
  >
    <h1 />
    <span>
      (undefined)
    </span>
  </button>
</div>
`;
